### ISSUE TITLE
Structures for not-just-lambda events?

### ISSUE THREAD

**Issue by JackKelly-Bellroy** [Nike-Inc/hal#137]
At work, we're considering using [Lambda Destinations](https://docs.aws.amazon.com/lambda/latest/dg/invocation-async.html#invocation-async-destinations) to respond to failed asynchronous lambda invocations. The obvious thing to do would be to PR some new types and `ToJSON`/`FromJSON` instances for that type, but they're not really lambda-specific: they can also be sent to SQS Queues, SNS Topics, or onto an EventBridge Event Bus. So there's a decent case for putting it in some other package

That got me thinking: should we look at separating `hal` from the event types? A package name like `aws-lambda-events` sounds like a fine home for events sent _to_ Lambda Functions (e.g., things that come from an Event Source Mapping) as well as events generated _by_ Lambda Functions (e.g., invocation records sent to a Lambda Destination, regardless of type). And the package would only really need to depend on `aeson` and testing libraries.

I'm interested to hear what you think here, and what namespace we could use to ensure a smooth transition. I expect we'd want to do something clever with re-exports or something so that existing code can keep working with warnings, and then migrate to using the new package as a direct dependency.

**Comment by IamfromSpace** [comment]
I agree that separating the two is probably a good idea.  The recent move of the Aeson parser was quite a pain for compatibility, so my lean is either to do this soon, or just commit to this being the long term place for events.  I think that's likely not a good idea because I just generally expect that to be a bit more volatile.

So, I think now is probably the best time if more schemas (events or destinations) are of interest.  Based on the Aeson changes, I think flags are the way to go.  It looks like we can add a flag to Hal that keeps event modules out so that they don't collide with the new package. It would default to the existing behavior today, and then maybe eventually we just no longer export them at all in future major version bump.

**Comment by JackKelly-Bellroy** [comment]
I'm not sure we need to do too much with flags to ensure a smooth experience for users.

What made things hard in the `aeson` case was IIRC that the package with no modules wasn't initially present, and its lower bounds excluded a lot of previous versions, and it needed `aeson` machinery in its implemenation. This forced downstream packages to use flags. I think we can avoid both of these if we're careful.

What I propose we do:

1. Decide on a namespace for AWS events that's not going to clash. `Network.AWS.*` is pretty full of amazonka <2.0 stuff, so my first thought of `Network.AWS.Events.*` is probably bad. Maybe `Data.AWS.Event.*`?
2. Decide on a package name for events. Maybe `aws-events`?
3. Decide which new version V of `hal` will include phase 1 of the migration
4. Move the event mapping structures across to `aws-events-0.1.0.0`. In version V of `hal`, re-export event mappings from `AWS.Lambda.Events.*` and mark these modules as deprecated.

I think this means packages can move across to `aws-events` at their own pace, and both can be present in the dependency graph without clashing.

**Comment by JackKelly-Bellroy** [comment]
We have some types for processing DynamoDB Stream events that we were thinking of upstreaming, but we'll need to do some work on them to make them fit your backwards-compatibility and dependency-footprint goals (namely, they currently depend on [`amazonka-dynamodb-streams`](https://hackage.haskell.org/package/amazonka-dynamodb-streams) to provide some of the internal data structures. While we're here, I was thinking it might be good to split off the existing events and get the AWS events package set up.

Before we went off and did this (in the space between our other work tasks), I just wanted to check in and see what capacity you had to look at and merge PRs, especially since reworking the package structure will be a bit of work.

Also, what are your current goals around GHC compatibility etc? The current release of `hal` declares a `base` bound that implies a GHC 7.something but your Stack configurations only go back to GHC 8.6.5 from 2019.

**Comment by IamfromSpace** [comment]
Apologies for the slow response.  Capacity is limited, but I'll do my best to make it work.  I definitely think we can relax compatibility some at this point too :)

This repo obviously hasn't been my main focus for a while, but I do want to keep things unblocked as much as I'm able.
